File: test/AdventureWorksLT/CustomerContactPerson.cs
using DevZest.Data;
using System;

namespace DevZest.Samples.AdventureWorksLT
{
    public static class CustomerContactPerson
    {
        private sealed class Manager : AttachedLocalColumnManager<Customer, string>
        {
            public static readonly Manager Singleton = new Manager();

            private Manager()
            {
            }

            protected override Column<string> CreateLocalColumn(DataSetContainer dataSetContainer, Customer _)
            {
                return dataSetContainer.CreateLocalColumn(_, _.LastName, _.FirstName, _.Title, GetContactPerson);
            }
        }

        public static Column<string> GetContactPerson(this Customer _)
        {
            return Manager.Singleton.GetAttachedColumn(_);
        }

        public static Action<Customer> Initializer
        {
            get { return Manager.Singleton.Initializer; }
        }

        private static string GetContactPerson(DataRow dataRow, _String lastName, _String firstName, _String title)
        {
            return GetContactPerson(lastName[dataRow], firstName[dataRow], title[dataRow]);
        }

        public static string GetContactPerson(string lastName, string firstName, string title)
        {
            string result = string.IsNullOrEmpty(lastName) ? string.Empty : lastName.ToUpper();
            if (!string.IsNullOrEmpty(firstName))
            {
                if (result.Length > 0)
                    result += ", ";
                result += firstName;
            }
            if (!string.IsNullOrEmpty(title))
            {
                result += " (";
                result += title;
                result += ")";
            }

            return result;
        }
    }
}

File: test/AdventureWorksLT.VB/CustomerContactPerson.vb
Imports DevZest.Data
Imports System.Runtime.CompilerServices

Namespace DevZest.Samples.AdventureWorksLT
    Module CustomerContactPerson
        Private NotInheritable Class Manager
            Inherits AttachedLocalColumnManager(Of Customer, String)

            Public Shared ReadOnly Singleton As Manager = New Manager()

            Private Sub New()
            End Sub

            Protected Overrides Function CreateLocalColumn(dataSetContainer As DataSetContainer, __ As Customer) As Column(Of String)
                Return dataSetContainer.CreateLocalColumn(__, __.LastName, __.FirstName, __.Title, AddressOf GetContactPerson)
            End Function
        End Class

        <Extension()>
        Function GetContactPerson(ByVal __ As Customer) As Column(Of String)
            Return Manager.Singleton.GetAttachedColumn(__)
        End Function

        Public ReadOnly Property Initializer As Action(Of Customer)
            Get
                Return Manager.Singleton.Initializer
            End Get
        End Property

        Private Function GetContactPerson(ByVal dataRow As DataRow, ByVal lastName As _String, ByVal firstName As _String, ByVal title As _String) As String
            Return GetContactPerson(lastName(dataRow), firstName(dataRow), title(dataRow))
        End Function

        Function GetContactPerson(ByVal lastName As String, ByVal firstName As String, ByVal title As String) As String
            Dim result As String = If(String.IsNullOrEmpty(lastName), String.Empty, lastName.ToUpper())

            If Not String.IsNullOrEmpty(firstName) Then
                If result.Length > 0 Then result += ", "
                result += firstName
            End If

            If Not String.IsNullOrEmpty(title) Then
                result += " ("
                result += title
                result += ")"
            End If

            Return result
        End Function
    End Module
End Namespace
